### DISCUSSION TITLE
Is there a way to know in the IDurableOrchestrationContext if the instance was already run

### DISCUSSION THREAD

**Discussion by jrodriguezax** [Azure/azure-functions-durable-extension#2479] [Q&A] [answered]
Hi,

I'm using IDurableOrchestrationClient.Restart with the parameter restartWithNewInstanceId = false to keep the same InstanceID.

What I need to know inside the orchestration context, is if the Instance is actually coming from that Restart call or is a brand new execution, is there a way to know that from within the Durable Orchestration Context?

**Comment by cgillum** [comment]
I'm curious to know what you would do if that information, if available, because that might depend on the specific strategy. But I think one way to know if it's a new instance or not is to simply check `context.IsReplaying` at the start of the orchestration. If it's `false`, then you know this is a new instance. Otherwise, it's still the old instance.

**Reply by jrodriguezax** [reply]
Hi! I appreciate your time to check this question, the reason I need to know that is to update some data I store about instances that should be retried. I have tested the context.IsReplaying but no matter if I call the StartAsync or RestartAsync (indicating to keep the same instance Id), the value is always false.

**Reply by cgillum** [reply]
Ah, that's expected if you're starting (or restarting) for the first time. I think I misread your question. There's unfortunately no way to distinguish between these two cases, I'm afraid. Now that I have more context, I'll post a separate answer with a separate suggestion.

**Comment by ltouro** [comment, marked as answer]
Suggestion 1: Create a Restart extension that use a new ID that is composed by the previous ID + more information. For instance, MyWorkflowId-Restarted. That would be pretty easy to check and to sanitize to the original ID. 

Suggestion 2: Make your orchestration function input to use a common interface that has `bool Restarted`. Create a Restart extension that cast the 'status.Input' to your common input interface and set it to True. Or you could make that an int and increment a counter. 

If you don't mind making your restart function aware the orchestration input class, that should be even easier.

Check the restart implementation:

![image](https://github.com/Azure/azure-functions-durable-extension/assets/5283919/ecd56849-9b7c-4892-835b-aaae9b7d309e)

**Reply by jrodriguezax** [reply]
Hi! Thanks for your suggestions! I did have thought about using an interface as the orchestration function input, so I think I will implement that. I was just trying to know if it was a way from within the orchestration context, but this was very helpful, thanks!

**Comment by cgillum** [comment]
There isn't a way from inside the orchestration to distinguish a new orchestration created for the first time or a recreated orchestration instance.

Another strategy to consider is whether you can solve your problem using `context.ContinueAsNew(...)` instead of `IDurableOrchestrationClient.Restart`. This has roughly the same effect as long as you don't need to change the instance ID.

To do this, you'd need to update the orchestration code to call `ContinueAsNew` if some condition is met. For example, you might be able to examine the orchestration input to know if an orchestration instance is using the "old" version that needs to be restarted and call `ContinueAsNew` only in that case. I don't know if we document this practice of using `ContinueAsNew` to update orchestrations, but it should be considered a recommended strategy, if it applies to your situation.

**Reply by jrodriguezax** [reply]
Hi! @cgillum I appreciate the time you spent looking into this question. I think I will implement one of the suggestions from Lucas Itouro, but I really appreciate your help! thank you!
